src/wh_server.c

```diff
@@ -53,26 +53,53 @@ static int _wh_Server_HandleCustomRequest(whServerContext* server,
 int wh_Server_Init(whServerContext* server, whServerConfig* config)
 {
     int rc = 0;
+
     if ((server == NULL) || (config == NULL)) {
         return WH_ERROR_BADARGS;
     }
 
     memset(server, 0, sizeof(*server));
-    if (
-        ((rc = wolfCrypt_Init()) == 0) &&
-        ((rc = wc_InitRng_ex(server->crypto->rng, NULL, INVALID_DEVID)) == 0)) {
-        rc = wh_Nvm_Init(server->nvm, config->nvm_config);
-        if (rc == 0) {
-            server->nvm->cb = config->nvm_config->cb;
-            server->nvm->context = config->nvm_config->context;
 
-            rc = wh_CommServer_Init(server->comm, config->comm_config);
-            if (rc == 0) {
-                /* All good */
-            }
+    rc = wolfCrypt_Init();
```

Comment:
Let's be a bit more explicit during initialization and tear down to make sure it is always safe to call Server_Cleanup:
1. Add int wc_inited to server_context.  0 means not inited yet, 1 = inited successfully and needs to be cleaned up in ServerCleanup.
2. Use context.dev_id == 0 to mean cryptocb not registered yet, == INVALID_DEVID for no cryptocb, and otherwise cryptocb registered.  Deregister the cryptocb in server_cleanup before wolfCrypt_Cleanup if not 0.  Note that invoking register/deregister will trigger events to be sent to the cryptocb handler to allow hw initialization late if the dev context is NULL when WOLF_CRYPTO_CB_CMD is set, which it should be.  see https://github.com/wolfSSL/wolfssl/blob/master/wolfcrypt/src/cryptocb.c#L334 for more info.
3. Not sure why we are init'ing the RNG, but if we need one per server context, then add a context member to track if the InitRNG was successful so we can FreeRNG in server_cleanup if necessary.
4. NVM Init now updates the nvm_context on init and correctly unsets it on init failure so no additional state is necessary here
5. We should update CommServer_init to ensure it can correctly cleanup on failure.

Reply:
on 2, I'm pretty sure 0 is a valid devId, we only reserve -2 as an invalid id. I think `wh_Server_Init` should take an explicit argument for devId the same way wolfCrypt does it and just require the caller to specify if they're using a devId or not to get around the memset 0 of the crypto struct.

Reply:
@billphipps can you clarify what you mean on 5? Action items for this PR of for a subsequent one? Since this PR just adds support for server CCBs I don't want to touch too much stuff